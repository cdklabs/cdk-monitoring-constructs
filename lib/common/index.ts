export * from "./alarm";
export * from "./metric";
export * from "./monitoring";
export * from "./url";
export * from "./widget";
export * from "./strings";
