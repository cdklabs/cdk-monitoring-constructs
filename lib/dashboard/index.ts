export * from "./BitmapDashboard";
export * from "./DashboardRenderingPreference";
export * from "./DynamicDashboardSegment";
export * from "./DashboardSegment";
export * from "./DashboardWithBitmapCopy";
export * from "./DefaultDashboardFactory";
export * from "./DynamicDashboardFactory";
export * from "./DefaultWidgetFactory";
export * from "./IDashboardFactory";
export * from "./IDynamicDashboardFactory";
export * from "./IWidgetFactory";
export * from "./MonitoringNamingStrategy";
export * from "./SingleWidgetDashboardSegment";
export * from "./widget";
