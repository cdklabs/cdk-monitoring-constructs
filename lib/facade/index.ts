export * from "./IMonitoringAspect";
export * from "./MonitoringFacade";
export * from "./MonitoringAspect";
