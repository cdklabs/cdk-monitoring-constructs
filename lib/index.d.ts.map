{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["index.ts"],"names":[],"mappings":"AAAA,cAAc,UAAU,CAAC;AACzB,cAAc,aAAa,CAAC;AAC5B,cAAc,UAAU,CAAC;AACzB,cAAc,cAAc,CAAC"}