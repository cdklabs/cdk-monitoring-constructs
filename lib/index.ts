export * from "./common";
export * from "./dashboard";
export * from "./facade";
export * from "./monitoring";
