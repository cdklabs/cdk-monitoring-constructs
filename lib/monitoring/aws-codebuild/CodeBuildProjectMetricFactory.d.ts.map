{"version":3,"file":"CodeBuildProjectMetricFactory.d.ts","sourceRoot":"","sources":["CodeBuildProjectMetricFactory.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,aAAa,EAAE,MAAM,4BAA4B,CAAC;AAC3D,OAAO,EAAE,QAAQ,EAAE,MAAM,2BAA2B,CAAC;AAErD,OAAO,EAAE,aAAa,EAAmB,MAAM,cAAc,CAAC;AAE9D,MAAM,WAAW,kCAAkC;IACjD,QAAQ,CAAC,OAAO,EAAE,QAAQ,CAAC;CAC5B;AAED,qBAAa,6BAA6B;IACxC,SAAS,CAAC,QAAQ,CAAC,aAAa,EAAE,aAAa,CAAC;IAChD,SAAS,CAAC,QAAQ,CAAC,aAAa,EAAE,aAAa,CAAC;IAChD,SAAS,CAAC,QAAQ,CAAC,OAAO,EAAE,QAAQ,CAAC;gBAGnC,aAAa,EAAE,aAAa,EAC5B,KAAK,EAAE,kCAAkC;IAS3C,gBAAgB;IAIhB,yBAAyB;IAIzB,sBAAsB;IAItB,qBAAqB;IAMrB,0BAA0B;IAS1B,0BAA0B;IAS1B,0BAA0B;CAQ3B"}