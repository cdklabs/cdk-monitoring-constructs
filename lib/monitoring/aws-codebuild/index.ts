export * from "./CodeBuildProjectMetricFactory";
export * from "./CodeBuildProjectMonitoring";
