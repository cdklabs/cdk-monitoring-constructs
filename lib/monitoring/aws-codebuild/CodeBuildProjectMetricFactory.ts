import { DimensionsMap } from "aws-cdk-lib/aws-cloudwatch";
import { IProject } from "aws-cdk-lib/aws-codebuild";

import {
  BaseMetricFactory,
  BaseMetricFactoryProps,
  MetricFactory,
  MetricStatistic,
} from "../../common";

export interface CodeBuildProjectMetricFactoryProps
  extends BaseMetricFactoryProps {
  readonly project: IProject;
}

export class CodeBuildProjectMetricFactory extends BaseMetricFactory<CodeBuildProjectMetricFactoryProps> {
  protected readonly dimensionsMap: DimensionsMap;
  protected readonly project: IProject;

  constructor(
    metricFactory: MetricFactory,
    props: CodeBuildProjectMetricFactoryProps
  ) {
    super(metricFactory, props);

    this.project = props.project;
    this.dimensionsMap = {
      ProjectName: props.project.projectName,
    };
  }

  metricBuildCount() {
    return this.metricFactory.adaptMetric(this.project.metricBuilds());
  }

  metricSucceededBuildCount() {
    return this.metricFactory.adaptMetric(this.project.metricSucceededBuilds());
  }

  metricFailedBuildCount() {
    return this.metricFactory.adaptMetric(this.project.metricFailedBuilds());
  }

  metricFailedBuildRate() {
    return this.metricFailedBuildCount().with({
      statistic: MetricStatistic.AVERAGE,
    });
  }

  metricDurationP99InSeconds() {
    return this.metricFactory.adaptMetric(
      this.project.metricDuration({
        label: "P99",
        statistic: MetricStatistic.P99,
      })
    );
  }

  metricDurationP90InSeconds() {
    return this.metricFactory.adaptMetric(
      this.project.metricDuration({
        label: "P90",
        statistic: MetricStatistic.P90,
      })
    );
  }

  metricDurationP50InSeconds() {
    return this.metricFactory.adaptMetric(
      this.project.metricDuration({
        label: "P50",
        statistic: MetricStatistic.P50,
      })
    );
  }
}
