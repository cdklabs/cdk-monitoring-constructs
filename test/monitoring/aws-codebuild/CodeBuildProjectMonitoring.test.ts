import { Duration, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { BuildSpec, Project } from "aws-cdk-lib/aws-codebuild";

import { AlarmWithAnnotation, CodeBuildProjectMonitoring } from "../../../lib";
import { addMonitoringDashboardsToStack } from "../../utils/SnapshotUtil";
import { TestMonitoringScope } from "../TestMonitoringScope";

test("snapshot test: all alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  const project = new Project(stack, "Project", {
    projectName: "TheProject",
    buildSpec: BuildSpec.fromObject({
      version: "0.2",
      phases: {
        build: {
          commands: ["echo hello"],
        },
      },
    }),
  });

  let numAlarmsCreated = 0;

  const monitoring = new CodeBuildProjectMonitoring(scope, {
    project,
    addDurationP99Alarm: {
      Warning: {
        maxDuration: Duration.minutes(10),
      },
    },
    addDurationP90Alarm: {
      Warning: {
        maxDuration: Duration.minutes(2),
      },
    },
    addDurationP50Alarm: {
      Warning: {
        maxDuration: Duration.minutes(1),
      },
    },
    addFailedBuildCountAlarm: {
      Warning: {
        maxErrorCount: 1,
      },
    },
    addFailedBuildRateAlarm: {
      Warning: {
        maxErrorRate: 0.1,
      },
    },
    useCreatedAlarms: {
      consume(alarms: AlarmWithAnnotation[]) {
        numAlarmsCreated = alarms.length;
      },
    },
  });

  addMonitoringDashboardsToStack(stack, monitoring);
  expect(numAlarmsCreated).toStrictEqual(5);
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
