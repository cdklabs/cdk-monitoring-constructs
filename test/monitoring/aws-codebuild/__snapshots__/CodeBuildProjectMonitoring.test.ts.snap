// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: all alarms 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Alarm7103F465": Object {
      "Properties": Object {
        "DashboardBody": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"widgets\\":[{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":4,\\"x\\":0,\\"y\\":0,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "ScopeTestTheProjectDurationP99WarningFE155920",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":4,\\"x\\":6,\\"y\\":0,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "ScopeTestTheProjectDurationP90Warning8DBB8532",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":4,\\"x\\":12,\\"y\\":0,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "ScopeTestTheProjectDurationP50Warning3EA02799",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":4,\\"x\\":18,\\"y\\":0,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "ScopeTestTheProjectFailureCountWarning0D41B9F5",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}},{\\"type\\":\\"metric\\",\\"width\\":6,\\"height\\":4,\\"x\\":0,\\"y\\":4,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"annotations\\":{\\"alarms\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "ScopeTestTheProjectFailureRateWarningD89D16B4",
                  "Arn",
                ],
              },
              "\\"]},\\"yAxis\\":{}}}]}",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
    "ProjectC78D97AD": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "NO_ARTIFACTS",
        },
        "Cache": Object {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:1.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "Name": "TheProject",
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "ProjectRole4CCB274E",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "{
  \\"version\\": \\"0.2\\",
  \\"phases\\": {
    \\"build\\": {
      \\"commands\\": [
        \\"echo hello\\"
      ]
    }
  }
}",
          "Type": "NO_SOURCE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "ProjectRole4CCB274E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ProjectRoleDefaultPolicy7F29461B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "ProjectC78D97AD",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "ProjectC78D97AD",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    Object {
                      "Ref": "ProjectC78D97AD",
                    },
                    "-*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ProjectRoleDefaultPolicy7F29461B",
        "Roles": Array [
          Object {
            "Ref": "ProjectRole4CCB274E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Resource": Object {
      "Properties": Object {
        "DashboardBody": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"widgets\\":[{\\"type\\":\\"text\\",\\"width\\":24,\\"height\\":1,\\"x\\":0,\\"y\\":0,\\"properties\\":{\\"markdown\\":\\"### CodeBuild Project **[TheProject](https://eu-west-1.console.aws.amazon.com/codebuild/home?region=eu-west-1#/projects/",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              ")**\\"}},{\\"type\\":\\"metric\\",\\"width\\":8,\\"height\\":5,\\"x\\":0,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Builds\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/CodeBuild\\",\\"Builds\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"stat\\":\\"Sum\\"}],[\\"AWS/CodeBuild\\",\\"SucceededBuilds\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"stat\\":\\"Sum\\"}],[\\"AWS/CodeBuild\\",\\"FailedBuilds\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"stat\\":\\"Sum\\"}]],\\"annotations\\":{\\"horizontal\\":[{\\"label\\":\\"FailedBuilds > 1 for 3 datapoints within 15 minutes\\",\\"value\\":1,\\"yAxis\\":\\"left\\"}]},\\"yAxis\\":{\\"left\\":{\\"min\\":0,\\"label\\":\\"Count\\",\\"showUnits\\":false}}}},{\\"type\\":\\"metric\\",\\"width\\":8,\\"height\\":5,\\"x\\":8,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Duration\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/CodeBuild\\",\\"Duration\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"label\\":\\"P50\\",\\"stat\\":\\"p50\\"}],[\\"AWS/CodeBuild\\",\\"Duration\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"label\\":\\"P90\\",\\"stat\\":\\"p90\\"}],[\\"AWS/CodeBuild\\",\\"Duration\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"label\\":\\"P99\\",\\"stat\\":\\"p99\\"}]],\\"annotations\\":{\\"horizontal\\":[{\\"label\\":\\"P99 > 600000 for 3 datapoints within 15 minutes\\",\\"value\\":600000,\\"yAxis\\":\\"left\\"},{\\"label\\":\\"P90 > 120000 for 3 datapoints within 15 minutes\\",\\"value\\":120000,\\"yAxis\\":\\"left\\"},{\\"label\\":\\"P50 > 60000 for 3 datapoints within 15 minutes\\",\\"value\\":60000,\\"yAxis\\":\\"left\\"}]},\\"yAxis\\":{\\"left\\":{\\"min\\":0,\\"label\\":\\"sec\\",\\"showUnits\\":false}}}},{\\"type\\":\\"metric\\",\\"width\\":8,\\"height\\":5,\\"x\\":16,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Failed Builds (rate)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/CodeBuild\\",\\"FailedBuilds\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\"]],\\"annotations\\":{\\"horizontal\\":[{\\"label\\":\\"FailedBuilds > 0.1 for 3 datapoints within 15 minutes\\",\\"value\\":0.1,\\"yAxis\\":\\"left\\"}]},\\"yAxis\\":{\\"left\\":{\\"min\\":0,\\"label\\":\\"Rate\\",\\"showUnits\\":false}}}}]}",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
    "ScopeTestTheProjectDurationP50Warning3EA02799": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P50 duration is too long.",
        "AlarmName": "Test-TheProject-Duration-P50-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P50",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ProjectName",
                    "Value": Object {
                      "Ref": "ProjectC78D97AD",
                    },
                  },
                ],
                "MetricName": "Duration",
                "Namespace": "AWS/CodeBuild",
              },
              "Period": 300,
              "Stat": "p50",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 60000,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestTheProjectDurationP90Warning8DBB8532": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P90 duration is too long.",
        "AlarmName": "Test-TheProject-Duration-P90-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P90",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ProjectName",
                    "Value": Object {
                      "Ref": "ProjectC78D97AD",
                    },
                  },
                ],
                "MetricName": "Duration",
                "Namespace": "AWS/CodeBuild",
              },
              "Period": 300,
              "Stat": "p90",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 120000,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestTheProjectDurationP99WarningFE155920": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P99 duration is too long.",
        "AlarmName": "Test-TheProject-Duration-P99-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P99",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ProjectName",
                    "Value": Object {
                      "Ref": "ProjectC78D97AD",
                    },
                  },
                ],
                "MetricName": "Duration",
                "Namespace": "AWS/CodeBuild",
              },
              "Period": 300,
              "Stat": "p99",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 600000,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestTheProjectFailureCountWarning0D41B9F5": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Failure count is too high.",
        "AlarmName": "Test-TheProject-Failure-Count-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "Dimensions": Array [
          Object {
            "Name": "ProjectName",
            "Value": Object {
              "Ref": "ProjectC78D97AD",
            },
          },
        ],
        "EvaluationPeriods": 3,
        "MetricName": "FailedBuilds",
        "Namespace": "AWS/CodeBuild",
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestTheProjectFailureRateWarningD89D16B4": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Failure rate is too high.",
        "AlarmName": "Test-TheProject-Failure-Rate-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "Dimensions": Array [
          Object {
            "Name": "ProjectName",
            "Value": Object {
              "Ref": "ProjectC78D97AD",
            },
          },
        ],
        "EvaluationPeriods": 3,
        "MetricName": "FailedBuilds",
        "Namespace": "AWS/CodeBuild",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 0.1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "Summary68521F81": Object {
      "Properties": Object {
        "DashboardBody": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"widgets\\":[{\\"type\\":\\"text\\",\\"width\\":24,\\"height\\":1,\\"x\\":0,\\"y\\":0,\\"properties\\":{\\"markdown\\":\\"### CodeBuild Project **[TheProject](https://eu-west-1.console.aws.amazon.com/codebuild/home?region=eu-west-1#/projects/",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              ")**\\"}},{\\"type\\":\\"metric\\",\\"width\\":8,\\"height\\":5,\\"x\\":0,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Builds\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/CodeBuild\\",\\"Builds\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"stat\\":\\"Sum\\"}],[\\"AWS/CodeBuild\\",\\"SucceededBuilds\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"stat\\":\\"Sum\\"}],[\\"AWS/CodeBuild\\",\\"FailedBuilds\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"stat\\":\\"Sum\\"}]],\\"annotations\\":{\\"horizontal\\":[{\\"label\\":\\"FailedBuilds > 1 for 3 datapoints within 15 minutes\\",\\"value\\":1,\\"yAxis\\":\\"left\\"}]},\\"yAxis\\":{\\"left\\":{\\"min\\":0,\\"label\\":\\"Count\\",\\"showUnits\\":false}}}},{\\"type\\":\\"metric\\",\\"width\\":8,\\"height\\":5,\\"x\\":8,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Duration\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/CodeBuild\\",\\"Duration\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"label\\":\\"P50\\",\\"stat\\":\\"p50\\"}],[\\"AWS/CodeBuild\\",\\"Duration\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"label\\":\\"P90\\",\\"stat\\":\\"p90\\"}],[\\"AWS/CodeBuild\\",\\"Duration\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\",{\\"label\\":\\"P99\\",\\"stat\\":\\"p99\\"}]],\\"annotations\\":{\\"horizontal\\":[{\\"label\\":\\"P99 > 600000 for 3 datapoints within 15 minutes\\",\\"value\\":600000,\\"yAxis\\":\\"left\\"},{\\"label\\":\\"P90 > 120000 for 3 datapoints within 15 minutes\\",\\"value\\":120000,\\"yAxis\\":\\"left\\"},{\\"label\\":\\"P50 > 60000 for 3 datapoints within 15 minutes\\",\\"value\\":60000,\\"yAxis\\":\\"left\\"}]},\\"yAxis\\":{\\"left\\":{\\"min\\":0,\\"label\\":\\"sec\\",\\"showUnits\\":false}}}},{\\"type\\":\\"metric\\",\\"width\\":8,\\"height\\":5,\\"x\\":16,\\"y\\":1,\\"properties\\":{\\"view\\":\\"timeSeries\\",\\"title\\":\\"Failed Builds (rate)\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"metrics\\":[[\\"AWS/CodeBuild\\",\\"FailedBuilds\\",\\"ProjectName\\",\\"",
              Object {
                "Ref": "ProjectC78D97AD",
              },
              "\\"]],\\"annotations\\":{\\"horizontal\\":[{\\"label\\":\\"FailedBuilds > 0.1 for 3 datapoints within 15 minutes\\",\\"value\\":0.1,\\"yAxis\\":\\"left\\"}]},\\"yAxis\\":{\\"left\\":{\\"min\\":0,\\"label\\":\\"Rate\\",\\"showUnits\\":false}}}}]}",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
